docs/man/man1/hal_input.1
.TH HAL_INPUT "1" "2007-02-25" "LinuxCNC Documentation" "HAL User's Manual"
.SH NAME
hal_input \- control HAL pins with any Linux input device, including USB HID devices
.SH SYNOPSIS
\fIloadusr\fR \fBhal_input\fR \fI[\-KRAL] inputspec ...\fR
.SH DESCRIPTION
hal_input is an interface between HAL and any Linux input device, including USB
HID devices.  For each device named, \fBhal_input\fR creates pins corresponding
to its keys, absolute axes, and LEDs.  At a fixed rate of approximately 10ms,
it synchronizes the device and the HAL pins.
.SH INPUT SPECIFICATION
The \fIinputspec\fR may be in one of several forms:
.TP
A string \fIS\fR
A substring or shell-style pattern match will be tested against the "name"
of the device, the "phys" (which gives information about how it is connected),
and the "id", which is a string of the form "Bus=... Vendor=... Product=...
Version=...".  You can view the name, phys, and id of attached devices by executing \fBless /proc/bus/input/devices\fR.  Examples:
.RS 12
.PD 0
SpaceBall
.PP
"Vendor=001f Product=0001"
.PP
serio*/input0
.RE
.PD
.TP  
A number \fIN\fR
This opens \fB/dev/input/event\fIN\fR.  Except for devices that are always
attached to the system, this number may change over reboots or when the device
is removed.  For this reason, using an integer is not recommended.
.PP
When several devices are identified by the same string, add "\fB:\fIN\fR" where
\fIN\fR is the index of the desired device.  For example, if \fBMouse\fR
matches \fBinput3\fR and \fBinput10\fR, then \fBMouse\fR and \fBMouse:0\fR
select \fBinput3\fR.  Specifying \fBmouse:1\fR selects \fRinput10\fR.
.PP
For devices that appear as multiple entries in /dev/input, these indices are
likely to stay the same every time.  For multiple identical devices, these
indices are likely to depend on the insertion order, but stay the same across
reboots as long as the devices are not moved to different ports or unplugged
while the machine is booted.
.PP
If the first character of the \fIinputspec\fR is a "+", then \fBhal_input\fR
requests exclusive access to the device.  The first device matching an
\fIinputspec\fR is used.  Any number of \fIinputspec\fRs may be used.
.PP
A \fIsubset option\fI may precede each \fIinputspec\fR.  The subset option
begins with a dash.  Each letter in the subset option specifies a device
feature to \fBinclude\fR.  Features that are not specified are excluded.
For instance, to export keyboard LEDs to HAL without exporting keys, use
.RS 12
hal_input \-L \fIkeyboard\fR ...
.RE

.SH DEVICE FEATURES SUPPORTED
.IP \(bu 4
EV_KEY (buttons and keys).  Subset \-K
.IP \(bu 4
EV_ABS (absolute analog inputs).  Subset \-A
.IP \(bu 4
EV_REL (relative analog inputs).  Subset \-R
.IP \(bu 4
EV_LED (LED outputs).  Subset \-L
.SH HAL PINS AND PARAMETERS
.SS For buttons
.TP
.B input.\fIN\fB.btn\-\fIname\fR bit out
.br
.ns
.TP
.B input.\fIN\fB.btn\-\fIname\fB\-not\fR bit out
Created for each button on the device.
.SS For keys
.TP
.B input.\fIN\fB.key\-\fIname\fB
.br
.ns
.TP
.B input.\fIN\fB.key\-\fIname\fB\-not
Created for each key on the device.
.SS For absolute axes
.TP
.B input.\fIN\fB.abs\-\fIname\fB\-counts\fR s32 out
.br
.ns
.TP
.B input.\fIN\fB.abs\-\fIname\fB\-position\fR float out
.br
.ns
.TP
.B input.\fIN\fB.abs\-\fIname\fB\-scale\fR parameter float rw
.br
.ns
.TP
.B input.\fIN\fB.abs\-\fIname\fB\-offset\fR parameter float rw
.br
.ns
.TP
.B input.\fIN\fB.abs\-\fIname\fB\-fuzz\fR parameter s32 rw
.br
.ns
.TP
.B input.\fIN\fB.abs\-\fIname\fB\-flat\fR parameter s32 rw
.br
.ns
.TP
.B input.\fIN\fB.abs\-\fIname\fB\-min\fR parameter s32 r
.br
.ns
.TP
.B input.\fIN\fB.abs\-\fIname\fB\-max\fR parameter s32 r
Created for each absolute axis on the device.  Device positions closer than
\fBflat\fR to \fBoffset\fR are reported as \fBoffset\fR in \fBcounts\fR, and
\fBcounts\fR does not change until the device position changes by at least
\fBfuzz\fR.  The position is computed as \fBposition\fR = (\fBcounts\fR \-
\fBoffset\fR) / \fBscale\fR.  The default value of \fBscale\fR and \fBoffset\fR
map the range of the axis reported by the operating system to [\-1,1].  The
default values of \fBfuzz\fR and \fBflat\fR are those reported by the operating
system.  The values of \fBmin\fR and \fBmax\fR are those reported by the
operating system.
.SS For relative axes
.TP
.B input.\fIN\fB.rel\-\fIname\fB\-counts\fR s32 out
.br
.ns
.TP
.B input.\fIN\fB.rel\-\fIname\fB\-position\fR float out
.br
.ns
.TP
.B input.\fIN\fB.rel\-\fIname\fB\-reset\fR bit in
.br
.ns
.TP
.B input.\fIN\fB.rel\-\fIname\fB\-scale\fR parameter float rw
.br
.ns
.TP
.B input.\fIN\fB.rel\-\fIname\fB\-absolute\fR parameter s32 rw
.br
.ns
.TP
.B input.\fIN\fB.rel\-\fIname\fB\-precision\fR parameter s32 rw
.br
.ns
.TP
.B input.\fIN\fB.rel\-\fIname\fB\-last\fR parameter s32 rw
Created for each relative axis on the device.  As long as \fBreset\fR is true,
\fBcounts\fR is reset to zero regardless of any past or current axis movement.
Otherwise, \fBcounts\fR increases or decreases according to the motion of the
axis.  \fBcounts\fR is divided by \fRposition\-scale\fR to give \fBposition\fR.
The default value of \fBposition\fR is 1.  There are some devices, notably
scroll wheels, which return signed values with less resolution than 32 bits.
The default value of \fBprecision\fR is 32.  \fBprecision\fR can be set to 8
for a device that returns signed 8 bit values, or any other value from 1 to 32.
\fBabsolute\fR, when set true, ignores duplicate events with the same value.
This allows for devices that repeat events without any user action to work
correctly.  \fBlast\fR shows the most recent count value returned by the
device, and is used in the implementation of \fBabsolute\fR.
.SS For LEDs
.TP
.B input.\fIN\fB.led\-\fIname\fR bit out
.br
.ns
.TP
.B input.\fIN\fB.led\-\fIname\fB\-invert\fR parameter bit rw
Created for each LED on the device.
.SH PERMISSIONS AND UDEV
By default, the input devices may not be accessible to regular
users--\fBhal_input\fR requires read-write access, even if the device has no
outputs.

Different versions of udev have slightly different, incompatible syntaxes.  For
this reason, it is not possible for this manual page to give an accurate
example.  The
.B udev(7)
manual page documents the syntax used on your Linux distribution.
To view it in a terminal, the command is
.BR "man 7 udev" .

.SH BUGS
The initial state of keys, buttons, and absolute axes are erroneously reported
as FALSE or 0 until an event is received for that key, button, or axis.
.SH SEE ALSO
\fBudev(8)\fR, \fBudev(7)\fR


docs/NEWS
check debian/changelog for recent news

# some of EMC2 History & News
#############################

2009.06.13 - mshaver added emc2/configs/smithy directory and related files
             for Smithy Company's line of cnc machines.

2006.11.17 - removed HAL types s8, u8, s16, and u16.  HAL now uses float
	     for analog values, bit for booleans, and s32 or u32 for
	     integer values.  The only users who will be affected are
	     those with Universal Stepper Controller cards.  See item 9 at
	     http://wiki.linuxcnc.org/cgi-bin/emcinfo.pl?UpdatingConfigurationsForDevelopmentVersions

2006.11.15 - merged classicladder 0.7.100

2006.11.04 - added a feed-hold HAL pin.  Added M codes to enable/disable
	     feed override, spindle override, feed hold, and adaptive feed
	     individually.  Feed override is M50, spindle override is M51,
	     adaptive feed is M52, and feed hold is M53.  G50/51 are no
	     longer used to enable/disable adaptive feed, use M52 instead.

2006.10.24 - added support for block delete as defined by Section 2.2.2
             in the RS274NGC_3 specs.

2006.10.24 - added support for optional program stop as defined by Section 
             2.2.3 in the RS274NGC_3 specs.

2006.10.14 - manpages have been created for most rtapi and hal calls

2006.10.01 - added a userspace simulator which includes rtapi threads.
	     Use "--enable-simulator" to enable this mode, which cannot
	     be used to control real hardware

2006.09.27 - added support to optionally save joint positions between
             runs ([TRAJ] POSITION_FILE = position.txt)

2006.09.26 - added support to specify UNITS in literal in the ini. 
             (e.g. [TRAJ] LINEAR_UNITS = mm )

2006.09.09 - fixed bug #1384883, optional stop was not optional.

2006.08.09 - AXIS is now a part of EMC2, not a separate source download.

2006.08.07 - Added a spindle-speed override. This allows to override the spindle
	     speed during a program run. The MIN_SPINDLE_OVERRIDE and 
	     MAX_SPINDLE_OVERRIDE in the ini assure that speed doesn't get 
	     reduced or increased too much. Both values are in %.

2006.08.06 - Documentation can be built from source with the 'make docs' target

2006.06.11 - added realtime support for jogwheels.  Three HAL pins per axis:
	     one accepts raw counts from the wheel, one sets the distance moved
	     per count, and one enables that axis to move.

2006.05.17 - added 'adaptive feed', a HAL pin that can be used to scale the
	     feedrate in real time.  Simular to feedrate override, but can
	     only vary from 0.0 to 1.0, and reacts much faster.  Intended 
	     for use with EDM and other applications that require closed 
	     loop control of feedrate.

2006.03.18 - A new pin "count" exists for freqgen.  It is the same as the
             internal symbol "count" shown in the hal introduction.pdf,
             or the result of up OR down if there was an "OR" component.

2006.03.16 - Modified the motion control module to behave more like other
             HAL modules - it no longer adds its functions to threads
             automatically.  All the sample configs have been changed to
             explicitly add the functions, users who have custom configs
             will need to do the same.

2006.03.12 - If a copy of axis 1.3a0 or newer is detected inside src/
             it will be automatically built and installed for
             --enable-run-in-place use

2006.03.10 - minor bugfix in tkemc for jogging

2006.03.10 - added genhexkins, with a minitetra sample config.
             This proves EMC2 works just as good with hexapods as EMC1 did.

2006.03.10 - added VTI support & sample configs

2006.03.07 - lots of fixes to the TP (most were wrong in the old one as well)

2006.03.05 - new Trajectory Planner

2006.02.20 - added a manpage for the runscript.

2006.02.12 - Fixed a bug that was causing the RT code to run very slowly,
             forcing users to set BASE_PERIOD very long.

2006.02.10 - The build process has been replaced, no more recursive $(MAKE)

2006.01.28 - fixed bug #1205237, mode switching sometimes hung

2006.01.08 - M62/M63 digital outputs synchronous to motion
             M64/M65 digital outputs that get set right away

2006.01.06 - work on halconfig.tcl, it is used to display/configure HAL data
             requires the BWidget package, realtime, and HAL to run.

2006.01.04 - fix bug #1387740, a motion that goes nowhere clears feedrate

2006.01.01 - the config picker has i18n support, a German translation exists

2005.12.20 - a new config picker has been written, if you start just the
             runscript this one helps you choose a config to run

2005.12.13 - new runscript, and config organization
             work towards the first release of EMC2

2005.xx.xx - EMC2 now supports tool-changing in complex scenarios

2005.xx.xx - classicladder now a HAL component, a software PLC

2004.xx.xx - HAL (the Hardware Abstraction Layer) is used for
	     addressing hardware drivers, and allows for simple configuration


#developers: this file should hold a history of improvements, bugfixes, etc
             it is useful for keeping track of development


src/hal/user_comps/hal_input.py
#!/usr/bin/env python3
#    Copyright 2007 Jeff Epler <jepler@unpythonic.net>
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

import linux_event
import sys
import os
import fcntl
import hal
import select
import time
import glob
import fnmatch
import select
from hal import *


def tohalname(s): return str(s).lower().replace("_", "-")


class HalWrapper:
    def __init__(self, comp):
        self._comp = comp
        self._drive = {}
        self._pins = set()
        self._params = set()

    def drive(self):
        for k, v in list(self._drive.items()):
            setattr(self._comp, k, v)

    def newpin(self, *args):
        self._pins.add(args[0])
        return self._comp.newpin(*args)

    def newparam(self, *args):
        self._params.add(args[0])
        return self._comp.newparam(*args)

    def __getitem__(self, k):
        if k in self._drive:
            return self._drive[k]
        return self._comp[k]

    def __setitem__(self, k, v):
        if k in self._params:
            self._comp[k] = v
            return
        if k in self._pins:
            self._comp[k] = v
            return
        raise KeyError(k)
        self._drive[k] = v


class HalInputDevice:
    def __init__(self, comp, idx, name, parts='KRAL'):
        self.device = linux_event.InputDevice(name)

        self.idx = idx
        self.codes = set()
        self.last = {}
        self.rel_items = []
        self.abs_items = []
        self.comp = comp
        self.parts = parts

        if 'K' in parts:
            for key in self.device.get_bits('EV_KEY'):
                key = tohalname(key)
                self.codes.add(key)
                comp.newpin("%s.%s" % (idx, key), HAL_BIT, HAL_OUT)
                comp.newpin("%s.%s-not" % (idx, key), HAL_BIT, HAL_OUT)
                self.set(key + "-not", 1)

        if 'R' in parts:
            for axis in self.device.get_bits('EV_REL'):
                name = tohalname(axis)
                self.codes.add(name)
                comp.newpin("%s.%s-position" % (idx, name), HAL_FLOAT, HAL_OUT)
                comp.newpin("%s.%s-counts" % (idx, name), HAL_S32, HAL_OUT)
                comp.newpin("%s.%s-reset" % (idx, name), HAL_BIT, HAL_IN)
                comp.newpin("%s.%s-scale" % (idx, name), HAL_FLOAT, HAL_IN)
                self.set(name + '-scale', 1.)
                self.rel_items.append(name)

        if 'A' in parts:
            for axis in self.device.get_bits('EV_ABS'):
                name = tohalname(axis)
                self.codes.add(name)
                absinfo = self.device.get_absinfo(axis)
                comp.newpin("%s.%s-position" % (idx, name), HAL_FLOAT, HAL_OUT)
                comp.newpin("%s.%s-counts" % (idx, name), HAL_S32, HAL_OUT)
                comp.newpin("%s.%s-is-pos" % (idx, name), HAL_BIT, HAL_OUT)
                comp.newpin("%s.%s-is-neg" % (idx, name), HAL_BIT, HAL_OUT)
                comp.newpin("%s.%s-scale" % (idx, name), HAL_FLOAT, HAL_IN)
                comp.newpin("%s.%s-offset" % (idx, name), HAL_FLOAT, HAL_IN)
                comp.newpin("%s.%s-fuzz" % (idx, name), HAL_S32, HAL_IN)
                comp.newpin("%s.%s-flat" % (idx, name), HAL_S32, HAL_IN)
                comp.newparam("%s.%s-min" % (idx, name), HAL_S32, HAL_RO)
                comp.newparam("%s.%s-max" % (idx, name), HAL_S32, HAL_RO)
                center = (absinfo.minimum + absinfo.maximum)/2.
                halfrange = (absinfo.maximum - absinfo.minimum)/2. or 1
                self.set(name + "-counts", absinfo.value)
                self.set(name + "-position",
                         (absinfo.value - center) / halfrange)
                self.set(name + "-scale", halfrange)
                self.set(name + "-offset", center)
                self.set(name + "-fuzz", absinfo.fuzz)
                self.set(name + "-flat", absinfo.flat)
                self.set(name + "-min", absinfo.minimum)
                self.set(name + "-max", absinfo.maximum)
                self.abs_items.append(name)

        self.ledmap = {}
        if 'L' in parts:
            for led in self.device.get_bits('EV_LED'):
                name = tohalname(led)
                self.ledmap[name] = led
                comp.newpin("%s.%s" % (idx, name), HAL_BIT, HAL_IN)
                comp.newpin("%s.%s-invert" % (idx, name), HAL_BIT, HAL_IN)
                self.last[name] = 0
                self.device.write_event('EV_LED', led, 0)

    def get(self, name):
        name = "%s.%s" % (self.idx, name)
        return self.comp[name]

    def set(self, name, value):
        name = "%s.%s" % (self.idx, name)
        self.comp[name] = value

    def update(self):
        while self.device.readable():
            ev = self.device.read_event()
            if ev.type == 'EV_SYN':
                continue
            elif ev.type == 'EV_SND':
                continue
            elif ev.type == 'EV_MSC':
                continue
            elif ev.type == 'EV_LED':
                continue
            elif ev.type == 'EV_KEY' and 'K' not in self.parts:
                continue
            elif ev.type == 'EV_REL' and 'R' not in self.parts:
                continue
            elif ev.type == 'EV_ABS' and 'A' not in self.parts:
                continue
            code = tohalname(ev.code)
            if code not in self.codes:
                print("Unexpected event", ev.type, ev.code, file=sys.stderr)
                continue
            if ev.type == 'EV_KEY':
                if ev.value:
                    self.set(code, 1)
                    self.set(code + "-not", 0)
                else:
                    self.set(code, 0)
                    self.set(code + "-not", 1)
            elif ev.type == 'EV_REL':
                self.set(code + "-counts",
                         self.get(code + "-counts") + ev.value)
            elif ev.type == 'EV_ABS':
                flat = self.get(code + "-flat")
                fuzz = self.get(code + "-fuzz")
                center = int(self.get(code + "-offset"))
                if ev.value < center-flat or ev.value > center+flat:
                    value = ev.value
                else:
                    value = center
                if abs(value - self.get(code + "-counts")) > fuzz:
                    self.set(code + "-counts", value)

        for a in self.abs_items:
            value = self.get(a + "-counts")
            scale = self.get(a + "-scale") or 1
            offset = self.get(a + "-offset")
            position = (value - offset) / scale
            self.set(a + "-position", position)
            # Use .01 because my Joystick isn't exactly zero at rest. maybe should be a parameter?
            self.set(a + "-is-neg", (position < -.01))
            self.set(a + "-is-pos", (position > .01))

        for r in self.rel_items:
            reset = self.get(r + "-reset")
            scale = self.get(r + "-scale") or 1
            if reset:
                self.set(r + "-counts", 0)
            self.set(r + "-position", self.get(r + "-counts") / scale)

        for k, v in list(self.last.items()):
            # Note: this is OK because the hal module always returns True or False for HAL_BIT values
            u = self.comp["%s.%s" % (
                self.idx, k)] != self.comp["%s.%s-invert" % (self.idx, k)]
            if u != self.last[k]:
                led = self.ledmap[k]
                self.device.write_event('EV_LED', led, u)
                self.last[k] = u


h = component("hal_input")
w = HalWrapper(h)
h.setprefix("input")
d = []
i = 0
parts = 'KRAL'
for f in sys.argv[1:]:
    if f.startswith("-"):
        parts = f[1:]
    else:
        try:
            d.append(HalInputDevice(w, i, f, parts))
        except LookupError as detail:
            raise SystemExit(detail)
        parts = 'KRAL'
        i += 1
w.drive()
h.ready()

fds = [dev.device.fileno() for dev in d]
try:
    while 1:
        select.select(fds, [], [], .01)
        for i in d:
            i.update()
        w.drive()
except KeyboardInterrupt:
    pass
